File: 10b.pl
#!/usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;


{   package SyntaxChecker;
    use Moo;
    has _stack => (is      => 'lazy',
                   builder => sub { [] },
                   clearer => '_clean_stack');

    my %score = (
        ')' => 1,
        ']' => 2,
        '}' => 3,
        '>' => 4,
    );
    my %expect = qw/ ( ) [ ] { } < > /;

    sub push_stack {
        my ($self, $ch) = @_;
        push @{ $self->_stack }, $ch;
    }

    sub pop_stack {
        my ($self) = @_;
        return pop @{ $self->_stack }
    }

    sub score {
        my ($self, $line) = @_;
        $self->_clean_stack;
        for my $ch (split //, $line) {
            if (exists $expect{$ch}) {
                $self->push_stack($expect{$ch});

            } elsif ($ch ne $self->pop_stack) {
                return 0
            }
        }
        my $score = 0;
        while (my $ch = $self->pop_stack) {
            $score *= 5;
            $score += $score{$ch};
        }
        return $score
    }
}

my @scores;
my $sc = 'SyntaxChecker'->new;
while (<>) {
    chomp;
    my $score = $sc->score($_);
    push @scores, $score if $score;
}
say +(sort { $a <=> $b } @scores)[@scores / 2];

__DATA__
[({(<(())[]>[[{[]{<()<>>
[(()[<>])]({[<{<<[]>>(
{([(<{}[<>[]}>{[]{[(<()>
(((({<>}<{<{<>}{[]{[]{}
[[<[([]))<([[{}[[()]]]
[{[{({}]{}}([{[{{{}}([]
{<[[]]>}<{[{[{[]{()[[[]
[<(<(<(<{}))><([]([]()
<{([([[(<>()){}]>(<<{{
<{([{{}}[<[[[<>{}]]]>[]]

File: 10b.py
#! /usr/bin/python3
import sys

class SyntaxChecker:
    stack = []

    scores = {')': 1, ']': 2, '}': 3, '>': 4}
    expect = {'(':')', '[':']', '{':'}', '<':'>'}

    def clean_stack (self):
        self.stack = []

    def push_stack (self, ch):
        self.stack.append(ch)

    def pop_stack (self):
        return self.stack.pop()

    def score (self, line):
        self.clean_stack()
        for ch in line:
            if ch in self.expect:
                self.push_stack(self.expect[ch])
            elif ch != self.pop_stack():
                return 0
        s = 0
        while len(self.stack):
            ch = self.pop_stack()
            s *= 5
            s += self.scores[ch]
        return s

scores = []
sc = SyntaxChecker()

with open(sys.argv[1], 'r') as fh:
    for line in fh:
        line = line[:-1]
        score = sc.score(line)
        if score:
            scores.append(score)
scores.sort()
print(scores[int(len(scores)/2)])
